src/materialized/src/prof.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use lazy_static::lazy_static;
+
+use jemalloc_ctl::raw;
+use std::os::unix::ffi::OsStrExt;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::{ffi::CString, time::Instant};
+use tempfile::NamedTempFile;
+
+#[derive(Copy, Clone, Debug)]
+pub enum ProfStartTime {
+    Instant(Instant),
+    TimeImmemorial,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct JemallocProfMetadata {
+    pub start_time: Option<ProfStartTime>,
+}
+
+#[derive(Debug)]
+// Per-process singleton object allowing control of jemalloc profiling facilities.
+pub struct JemallocProfCtl {
+    md: JemallocProfMetadata,
+}
+
+#[cfg(target_os = "macos")]
+impl JemallocProfCtl {
+    fn get() -> Option<Self> {
+        None
+    }
+    pub fn get_md(&self) -> JemallocProfMetadata {
+        unreachable!()
+    }
+
+    pub fn activate(&mut self) -> Result<(), jemalloc_ctl::Error> {
+        unreachable!()
+    }
+
+    pub fn deactivate(&mut self) -> Result<(), jemalloc_ctl::Error> {
+        unreachable!()
+    }
+
+    pub fn dump(&mut self) -> anyhow::Result<std::fs::File> {
+        unreachable!()
+    }
+}
```

Comment:
It would be better to not implement this at all if the target os is not macos, and have different cfg-d functions in the http layer, as I mention below. That makes it impossible to accidentally crash materialize because you configure it incorrectly.

Alternatively, these could all return `Err` instead of panicking, but that is a bit harder to reason about.

Reply:
> That makes it impossible to accidentally crash materialize because you configure it incorrectly.

That's already not possible. `JemallocProfCtl` is only ever constructed by calling `get`, which in this impl returns `None`. And someone can't accidentally construct it from outside the module and violate that invariant, because it has private fields.

Actually I think this file (or crate, whatever) is the right place for the cfg logic to live. If we do it in `http`, we would have to handle two cases separately: (1) running on macOS (detected via cfg flags), and (2) running on Linux, but not having enabled profiling at runtime. Also, we would have to do the same logic in any future crate or module that depends on this one. I think the `prof` crate is the right place to encapsulate all logic related to checking whether memory profiling is possible.

Reply:
> That's already not possible. JemallocProfCtl is only ever constructed by calling get, which in this impl returns None.

Hmmm, I see your point, that assuages my concerns.

> Actually I think this file (or crate, whatever) is the right place for the cfg logic to live.

Yeah I think it would be best if prof crate always works "correctly" even if correctly means never working in some situations. That said, if it's impossible for the prof code to work in an environment then folks will get an earlier error (compile-time, instead of runtime) if the symbols literally don't exist on OSs that profiling information can't be generated for.

Put another way, the http code needs to have an `if` in some format, it's either a dynamic conditional or a compile-time `cfg`.

Like I said top level though, you do you: this code is not broken I'm just advocating for the way I'd have written it with my rationale so you can decide for yourself if it's persuasive.